### ISSUE TITLE
Considerations about the URLs

### ISSUE THREAD

**Issue by jancborchardt** [nextcloud/spreed#222]
So @LukasReschke and me talked about which improvements we should bring to the way our URLs look.

Currently they are like this:

`https://cloud.example.com/apps/spreed/?roomId=2`

Simply counting up the ID as there are more rooms. This has a few drawbacks: Can be guessed, doesn’t take into account when there is a custom name given, simply looks strange & a bit amateurish, and it’s not even that easily dictatable because there’s the technical `?roomId=` part.

So what about this. By default the URL looks something like this:

`https://cloud.example.com/apps/spreed/e2a5`

- A random 4-character string, with only lowercase characters and numbers.
- From the set of readable characters, meaning it does not have l, 1, I (lowercase L, number one, uppercase i) or O 0 (uppercase o, number 0) etc. to avoid confusion. @LukasReschke mentioned there is a list/library in PHP already.
- If the 4 characters are depleted, just start adding a fifth one. And so on …
- It’s dictatable, without any »code-looking« stuff, except the domain and slashes.
- Then when you rename the call and give it a custom name, that name should be used as ID. For example `https://cloud.example.com/apps/spreed/fosdem-planning` (lowercased, spaces replaced by a dash `-`). That allows it to be even more easily shareable/dictatable, which for a shared call is one of the design goals.

What do you think @nextcloud/spreed?

**Comment by sushidave** [comment]
Yes, do it simple! How about this: https://cloud.example.com/call/e2a5

**Comment by nickvergessen** [comment]
Fine by me, what about the case when two calls with custom name "Company call" exist? `company-call` and `company-call-2`? (I see a privacy issue here...)

**Comment by jancborchardt** [comment]
@nickvergessen it’s not any worse than the enumeration we do now, so I don’t see an issue.

@sushidave as far as I know this is not possible because the app lives in the apps/spreed folder. Or is it @Ivansss @nickvergessen?

**Comment by nickvergessen** [comment]
> it’s not any worse than the enumeration we do now, so I don’t see an issue.

The current enumeration has no "content" information. With named links if you create a room "i-hate-the-new-spreed-app" and you get the url "i-hate-the-new-spreed-app-2", you know that others already created such a room. Whoever if your room number is 42, yeah the other 41 rooms could be about anything.

**Comment by jancborchardt** [comment]
@LukasReschke what do you think?

I tend to say let's just go with it. The current situation is really strange.

**Comment by MorrisJobke** [comment]
> @LukasReschke what do you think?
> 
> I tend to say let's just go with it. The current situation is really strange.

Same goes for our current numeric ids ... and it only works for public rooms anyway because you need to be invited to private rooms. I'm fine with that guessing.

**Comment by nickvergessen** [comment]
Could also do `call/<id>-room-name` would prevent the duplicate issue and also when you later rename the call, old links would still work since the id remains part of the url

**Comment by nickvergessen** [comment]
Ah, now i remember the issue why we started using `?roomId=` in first place. The browser history feature can only save parameters. But we can just do the same as with files and forward the call/... to the `?roomId=` page. So entry is easy, but navigating back and forth is also painless

**Comment by rugk** [comment]
Don't want to get into [room 101](https://en.wikipedia.org/wiki/Ministries_of_Nineteen_Eighty-Four#Room_101), so please change the URLs… :wink: 

Seriously I'd also say `call/<id>-room-name` is the best way. As for the browser history you should be able to manipulate it with the [history API](https://developer.mozilla.org/en-US/docs/Web/API/History_API). However, of course, you can also use `call/?<id>-room-name` or so, the ? would hardly matter.

However I would at least propose to use at least 8-12 characters by default and possibly mix uppercase and lowercase. As a [shortlink study](https://arxiv.org/pdf/1604.02734v1.pdf) showed these URLs could otherwise be brute-forced otherwise. As [the authors explained](https://freedom-to-tinker.com/2016/04/14/gone-in-six-characters-short-urls-considered-harmful-for-cloud-services/) Google changed their short link generator to use "11- or 12-character tokens".

**Comment by nickvergessen** [comment]
Well since for our links you still need to be invited into the room, this does not really matter a lot.
But yeah it's still on my todo

**Comment by rugk** [comment]
How do you need to be invited? You get a link, access it and you're in AFAIK…

Of course one good thing is that 1) there are multiple Nextcloud servers, so you don't know all of them and 2) you'll see when someone enters the "brute force problem" is mitigated, but anyway wqhen one knows the NextCloud url they could brute-force all rooms to get in.

**Comment by nickvergessen** [comment]
Yeah but that only works for public rooms, and there doesn't even have to be one. Also with any new urls thw name will also be ignored and only the number is important, otherwise you would need to send out new links when you rename a room...

**Comment by rugk** [comment]
Okay, so maybe the URL should be like this:
`/call/?RoomId#cool-room-name`

The name would not matter in this case and could just be ignored. A plus punt would be that the name us not sent to the server, so e.g. in the sever logs it does nor appear. This reduces the meta data, which I think is an important thing nowadays.
Also hash anchors are easy to rename and read in JS and they don't affect the browser history.

As for the brute force problem the security would depend solely on the room ID, which is okay. However maybe let the sever admin configure how long the ID should be, i.e. whether easy-to-remember is preferred or security.

**Comment by rugk** [comment]
As for "public rooms" I did not even know there are private ones. As far as I see I can just share the link to anyone and they can enter into the room. When is a room private?

**Comment by nickvergessen** [comment]
When a room is not shared, it can only be accessed by the internal users that have been added to the room.

**Comment by rugk** [comment]
Ah okay so I have to click on the link for some users. I think the UI could be improved here. E.g. when I first used it, I thought I could only invite server users and there is no public chat at all. The rain was that u found no "invite other people" or "copy link" button.
However I found such a button when adding a sever user, but as the link was connected to the sever user in the UI I thought only that user could/should open the link or so. But am I right that this is the link I can share to other not registered people?
